{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0",
   "metadata": {},
   "outputs": [],
   "source": [
    "source(paste0(dirname(getwd()),'/map.r'))\n",
    "source(paste0(HELP_DIR, \"shortcuts.r\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1",
   "metadata": {},
   "source": [
    "#### 0 - Read somatic exome and RNA VAF\n",
    "- RNA Vafs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2",
   "metadata": {},
   "outputs": [],
   "source": [
    "rna_vaf <- fread(paste0(TMP_DIR, \"rna_vaf.csv\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3",
   "metadata": {},
   "source": [
    "- Directory with whole somatic exome output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4",
   "metadata": {},
   "outputs": [],
   "source": [
    "EXOME_DIR <- paste0(TMP_DIR, \"somatic_exome_big/\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5",
   "metadata": {},
   "source": [
    "- Remove non-translated reads (make much smaller, and easier to work with)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6",
   "metadata": {},
   "outputs": [],
   "source": [
    "remove <- c(\"3_prime_UTR_variant\",\"5_prime_UTR_variant\",\"intron_variant\",\"upstream_gene_variant\",\"non_coding_transcript_exon_variant\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7",
   "metadata": {},
   "source": [
    "### 1 - Run the joins"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8",
   "metadata": {},
   "outputs": [],
   "source": [
    "go <- data.frame()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9",
   "metadata": {},
   "outputs": [],
   "source": [
    "for( i in list.files(EXOME_DIR)){\n",
    "  fp <- paste0(EXOME_DIR, i)\n",
    "  print(fp); flush.console()\n",
    "  tmp <- \n",
    "  fread( fp ) %>% \n",
    "    fi(! annotation %in% remove) %>% \n",
    "    lj(rna_vaf, by = c(\"sampleId\", \"chromosome\", \"position\", \"REF\", \"ALT\", \"gene\", \"transcript\"))\n",
    "  go <- rbind(go, tmp)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "10",
   "metadata": {},
   "source": [
    "### 2 - Write output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "11",
   "metadata": {},
   "outputs": [],
   "source": [
    "fwrite(go, paste0(TMP_DIR, \"somatic_exome.csv\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "12",
   "metadata": {},
   "outputs": [],
   "source": [
    "df <- fread(paste0(TMP_DIR, \"drivers.csv\")) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "13",
   "metadata": {},
   "outputs": [],
   "source": [
    "fread(\"/mnt/bioinfnas2/immunocomp/shared_reference_data/chatGPT_gene_annotation.csv\") %>% head(12)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "14",
   "metadata": {},
   "outputs": [],
   "source": [
    "fread(paste0(TMP_DIR, \"drivers.csv\")) %>% \n",
    " gb(likelihoodMethod) %>% \n",
    " su(ct = n()) %>% \n",
    " ar(desc(ct))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "15",
   "metadata": {},
   "outputs": [],
   "source": [
    "fread(paste0(TMP_DIR, \"structural_variants/breakend.csv\")) %>% \n",
    " fi(disruptive) %>% \n",
    " se(sampleId, svId, gene, disruptive, reportedDisruption, undisruptedCopyNumber) \n",
    " %>% unique() "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
