{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import gzip\n",
    "from helpers import *"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1",
   "metadata": {},
   "source": [
    "### Helpers "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2",
   "metadata": {},
   "source": [
    "- Old"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3",
   "metadata": {},
   "source": [
    "- Newer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Functions to Read VCF file ###\n",
    "def get_vcf_col_names(vcf_path: str) -> list:\n",
    "    with gzip.open(vcf_path, \"rt\") as ifile:\n",
    "          for line in ifile:\n",
    "            if line.startswith(\"#CHROM\"):\n",
    "                  vcf_names = [x for x in line.split('\\t')]\n",
    "                  break\n",
    "    ifile.close()\n",
    "    return vcf_names\n",
    "\n",
    "def read_vcf(vcf_path: str) -> pd.DataFrame:\n",
    "    names = get_vcf_col_names(vcf_path)\n",
    "    return pd.read_csv(vcf_path, compression='gzip', comment='#', chunksize=10000, sep='\\s+', header=None, names=names).read()\n",
    "\n",
    "### Metadata ###\n",
    "def get_metadata_df(vcf: pd.DataFrame) -> pd.DataFrame:\n",
    "    return vcf[[\"#CHROM\", \"POS\",  \"REF\", \"ALT\", \"FILTER\"]].rename(columns = {'#CHROM':'chromosome', \"POS\" : \"position\"})\n",
    "\n",
    "#### pass INFO column #### \n",
    "def get_impact_fields( s: str ) :\n",
    "    impact_fields = s.split(\"IMPACT=\")[1].split(\";\")[0].split(\",\")\n",
    "    return impact_fields    \n",
    "        \n",
    "def get_impact_field(s: str, field: str):\n",
    "    if 'IMPACT=' in s: \n",
    "        impact = get_impact_fields( s )\n",
    "        if field == 'gene':\n",
    "            return impact[0]\n",
    "        elif field == 'transcript':\n",
    "            return impact[1]\n",
    "        elif field == 'impact':\n",
    "            return impact[2]\n",
    "    else:\n",
    "        return pd.NA \n",
    "        \n",
    "def extract_info_fields(s: str) -> dict:\n",
    "    return {\n",
    "        'gene': get_impact_field(s, \"gene\"), \n",
    "        'transcript' : get_impact_field(s, \"transcript\"), \n",
    "        'impact': get_impact_field(s, \"impact\")\n",
    "     }    \n",
    "\n",
    "def get_info_df(vcf: pd.Series) -> pd.DataFrame:\n",
    "    return pd.DataFrame([extract_info_fields(i) for i in vcf['INFO']])\n",
    "\n",
    "### Extra RNA and DNA depth/vaf ### \n",
    "def get_rna_dna_dp_vaf(sample: str, vcf: pd.DataFrame) -> pd.DataFrame:\n",
    "    return pd.DataFrame(\n",
    "    {'dna_vaf' : [i.split(\":\")[3] for i in vcf[sample]],\n",
    "     'dna_dp' : [i.split(\":\")[4] for i in vcf[sample]],\n",
    "     'rna_vaf' : [i.split(\":\")[3] for i in vcf[sample + '_RNA\\n']],\n",
    "     'rna_dp' : [i.split(\":\")[4] for i in vcf[sample + '_RNA\\n']]\n",
    "    })\n",
    "\n",
    "### Together ### \n",
    "def vamos(fp: str) -> pd.DataFrame:\n",
    "    sample = fp.split(\".\")[0]\n",
    "    try:\n",
    "        if os.path.exists(fp):\n",
    "            vcf = read_vcf(fp)\n",
    "            metadata = get_metadata_df( vcf )\n",
    "            info = get_info_df( vcf )\n",
    "            vaf = get_rna_dna_dp_vaf( sample, vcf)\n",
    "            together = metadata.join(info).join(vaf)\n",
    "            ready = together.query('FILTER==\"PASS\" & gene != \"\")')    \n",
    "            ready[['sampleId']] = sample\n",
    "        return ready\n",
    "    except Exception as e:\n",
    "        print(f\"Unexpected error: {e}\")\n",
    "        return None                 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5",
   "metadata": {},
   "outputs": [],
   "source": [
    "fp = vafs[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6",
   "metadata": {},
   "outputs": [],
   "source": [
    "together"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7",
   "metadata": {},
   "outputs": [],
   "source": [
    "#vamos(fp)\n",
    "sample = fp.split(\".\")[0]\n",
    "vcf = read_vcf(fp)\n",
    "metadata = get_metadata_df( vcf )\n",
    "info = get_info_df( vcf )\n",
    "vaf = get_rna_dna_dp_vaf( sample, vcf)\n",
    "df = metadata.join(info).join(vaf)\n",
    "\n",
    "df_go = df[(df['gene'].notna()) & (df['FILTER'] == \"PASS\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_go[df_go['impact'] == \"missense_variant\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9",
   "metadata": {},
   "source": [
    "#### 0 - Go!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "10",
   "metadata": {},
   "outputs": [],
   "source": [
    "os.chdir(SAGE_DIR)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "11",
   "metadata": {},
   "outputs": [],
   "source": [
    "vafs = [i for i in os.listdir() if 'purple.somatic' in i]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "12",
   "metadata": {},
   "outputs": [],
   "source": [
    "vaf_dfs = []\n",
    "for f in vafs[:3]:\n",
    "    print(\"Processing \" + f)\n",
    "    vaf_dfs.append( go(f) )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "13",
   "metadata": {},
   "source": [
    "#### 1 - Pre-filtering of the data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "14",
   "metadata": {},
   "outputs": [],
   "source": [
    "tmp = pd.concat(vaf_dfs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "15",
   "metadata": {},
   "outputs": [],
   "source": [
    "tmp['rna_dp'] = tmp['rna_dp'].astype(int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "16",
   "metadata": {},
   "outputs": [],
   "source": [
    "tmp.to_csv(READY_DIR + \"vaf_test.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "17",
   "metadata": {},
   "outputs": [],
   "source": [
    "READY_DIR + \"vaf_test.csv\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "18",
   "metadata": {},
   "outputs": [],
   "source": [
    "def filter_by_column(df, column, value):\n",
    "    return df[df[column] > value]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "19",
   "metadata": {},
   "outputs": [],
   "source": [
    "filtered_df = tmp.pipe(filter_by_column, 'rna_dp', 4)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "20",
   "metadata": {},
   "source": [
    "#### 2 - Send it!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "21",
   "metadata": {},
   "outputs": [],
   "source": [
    "filtered_df.to_csv( READY_DIR + \"vaf_ready.csv\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
