{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0",
   "metadata": {},
   "outputs": [],
   "source": [
    "library(data.table)\n",
    "library(tidyverse)\n",
    "library(gridExtra)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1",
   "metadata": {},
   "source": [
    "### 0 - Data Ready"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2",
   "metadata": {},
   "outputs": [],
   "source": [
    "vafs <- fread(\"/home/jusset/vaf_clean/data/vaf_ready.csv\") %>% select(-V1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3",
   "metadata": {},
   "outputs": [],
   "source": [
    "saveRDS(vafs %>% filter(tier == \"PANEL\") %>% arrange(gene) %>% pull(gene) %>% unique(), \n",
    "        file = \"/home/jusset/vaf_clean/data/panel_gene.Rds\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4",
   "metadata": {},
   "source": [
    "### 1 - Prep RNA VAF\n",
    "- Remove subclonal variants\n",
    "- Require purple_cn to be greater than 1\n",
    "- Threshold purple allele frequency and RNA VAF estimate to be max 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5",
   "metadata": {},
   "outputs": [],
   "source": [
    "vaf_ready <- \n",
    "vafs %>% \n",
    "  filter( subclonal < .1, purple_cn > .5, !is.na(purple_af) ) %>% \n",
    "  mutate( purple_af_mx = ifelse(biallelic, 1, purple_af)) %>% #### changed to biallelic\n",
    "  mutate( rna_vaf_est = ifelse(rna_vaf/purple_af_mx > 1, 1, rna_vaf/purple_af_mx))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6",
   "metadata": {},
   "source": [
    "### 2 - Aggregations at Gene-level\n",
    "- Aggregate sample and gene level aggregations "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7",
   "metadata": {},
   "outputs": [],
   "source": [
    "lets_go <- \n",
    "vaf_ready %>%\n",
    "  group_by(sample, gene) %>% \n",
    "  summarise(\n",
    "    rna_vaf = sum(rna_vaf_est*rna_dp)/sum(rna_dp),\n",
    "    rna_dp = sum(rna_dp),\n",
    "    rna_dp_adj = sum(rna_dp * purple_af_mx),\n",
    "    purple_af_mn = mean(purple_af_mx),\n",
    "    vafs = n(),\n",
    "    biallelic = paste0(unique(biallelic), collapse = \",\"), \n",
    "    afs = paste0(purple_af, collapse = \",\")  \n",
    "  ) %>% ungroup() "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8",
   "metadata": {},
   "source": [
    "### 3 - Send it"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9",
   "metadata": {},
   "outputs": [],
   "source": [
    "fwrite(lets_go, \"/home/jusset/vaf_clean/data/vaf_ready.txt\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
