{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0",
   "metadata": {},
   "outputs": [],
   "source": [
    "source(paste0(dirname(dirname(getwd())),'/map.r'))\n",
    "source(paste0(HELP_DIR, \"shortcuts.r\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1",
   "metadata": {},
   "source": [
    "# 0 - Read somatic exome data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2",
   "metadata": {},
   "outputs": [],
   "source": [
    "somatic_exome <- \n",
    "fread(\"/mnt/petasan_immunocomp/datasets/hartwig/biomarkers/database/somatic_exome.csv\") %>% \n",
    " gb(chromosome, position, gene, tier) %>% \n",
    " mu(ct = n(), hotspot = (tier == \"HOTSPOT\" | ct > 30)) %>% \n",
    " ug()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3",
   "metadata": {},
   "outputs": [],
   "source": [
    "annotater <- function(i){\n",
    " if(grepl(\"stop\", i)){ \"nonsense\" }\n",
    " else if(grepl(\"missense\", i)){ \"missense\" }       \n",
    " else if(grepl(\"inframe\", i)){ \"indel\" }\n",
    " else if(grepl(\"frameshift\", i)){ \"indel\" }    \n",
    " else if(grepl(\"start_lost\", i)){ \"nonsense\" }\n",
    " else {\"synonymous\"}\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4",
   "metadata": {},
   "outputs": [],
   "source": [
    "somatic_exome <- \n",
    "somatic_exome %>% \n",
    " rw() %>% mu( type = annotater(annotation)) %>% ug() %>%\n",
    " fi(type != \"synonymous\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5",
   "metadata": {},
   "source": [
    "# 1 - Compute hotspots, biallelic, inframe"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6",
   "metadata": {},
   "source": [
    "- Hotspots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7",
   "metadata": {},
   "outputs": [],
   "source": [
    "hotspots <- \n",
    "somatic_exome %>% \n",
    " fi(hotspot) %>% \n",
    " tm(sampleId, chromosome, gene, transcript, annotation, biallelic, method = \"HOTSPOT\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8",
   "metadata": {},
   "source": [
    "- Biallelic"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9",
   "metadata": {},
   "outputs": [],
   "source": [
    "biallelic <- \n",
    "somatic_exome %>% \n",
    " fi(!hotspot, biallelic) %>% \n",
    " tm(sampleId, chromosome, gene, transcript, annotation, biallelic, method = \"BIALLELIC\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "10",
   "metadata": {},
   "source": [
    "- Inframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "11",
   "metadata": {},
   "outputs": [],
   "source": [
    "inframe <- \n",
    "somatic_exome %>% \n",
    " fi(!hotspot, !biallelic, type == \"indel\") %>% \n",
    " tm(sampleId, chromosome, gene, transcript, annotation, biallelic, method = \"INFRAME\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "12",
   "metadata": {},
   "source": [
    "- Somatic drivers "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "13",
   "metadata": {},
   "outputs": [],
   "source": [
    "#som_drivers <- rbind(hotspots, biallelic, inframe) %>% lj(cn_gene %>% se(gene, chromosomeBand), by = \"gene\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "14",
   "metadata": {},
   "source": [
    "# 2 - Now DNDS! "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "15",
   "metadata": {},
   "source": [
    "- Read in DNDS output, create referencd for labelling DNDS drivers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "16",
   "metadata": {},
   "outputs": [],
   "source": [
    "sel_cvs <- fread(\"/mnt/petasan_immunocomp/datasets/hartwig/biomarkers/database/dnds/sel_cvs.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "17",
   "metadata": {},
   "outputs": [],
   "source": [
    "w_thresh <- 2\n",
    "q_thresh <- .05"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "18",
   "metadata": {},
   "outputs": [],
   "source": [
    "gper <- function(i){\n",
    " if(grepl(\"non\", i)){ \"nonsense\" }\n",
    " else if(grepl(\"mis\", i)){ \"missense\" }       \n",
    " else { \"indel\" }\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "19",
   "metadata": {},
   "outputs": [],
   "source": [
    "dnds_drivers_ref <- \n",
    "sel_cvs %>% \n",
    " fi(grepl(\"Pan-Cancer\", cohort), \n",
    "    (qtrunc_cv < q_thresh | qmis_cv < q_thresh | qind_cv < q_thresh),\n",
    "    (wmis_cv > w_thresh | wnon_cv > w_thresh | wind_cv > w_thresh )) %>% \n",
    " mu(tot = n_syn\t+ n_mis\t+ n_non + n_spl\t+ n_ind) %>% \n",
    " se(gene_name, tot, wmis_cv, wnon_cv, wind_cv, qmis_cv, qtrunc_cv, qind_cv) %>% \n",
    " ga( w, wval, -gene_name, -tot, -qmis_cv, -qtrunc_cv, -qind_cv) %>% \n",
    " ga( q, qval, -gene_name, -tot, -w, -wval) %>% \n",
    " fi(wval > w_thresh, qval < q_thresh, \n",
    "   (grepl(\"mis\", w) & grepl(\"mis\", q)) | (grepl(\"ind\", w) & grepl(\"ind\", q)) | (grepl(\"non\", w) & grepl(\"trunc\", q))) %>%\n",
    " rw() %>% mu(type = gper(w)) %>% ug() %>% \n",
    " tm(gene = gene_name, type)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "20",
   "metadata": {},
   "outputs": [],
   "source": [
    "sel_cvs %>% fi(gene_name == \"CSMD1\") %>% fi(cohort == \"Pan-Cancer\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "21",
   "metadata": {},
   "source": [
    "- Annotate variant types in somatic exome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "22",
   "metadata": {},
   "outputs": [],
   "source": [
    "dnds <- \n",
    "somatic_exome %>% \n",
    " fi( gene %in% unique(dnds_drivers_ref$gene)) %>% \n",
    " se( sampleId, chromosome, gene, transcript, type, biallelic ) %>% \n",
    " ij( dnds_drivers_ref, by = c(\"gene\", \"type\")) %>% \n",
    " mu( method = \"DNDS\" ) %>% \n",
    " rename(annotation = type)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "23",
   "metadata": {},
   "source": [
    "# 2 - Together"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "24",
   "metadata": {},
   "outputs": [],
   "source": [
    "somatic_drivers <- rbind(dnds, hotspots, biallelic, inframe) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "25",
   "metadata": {},
   "outputs": [],
   "source": [
    "fwrite(somatic_drivers, \"/mnt/petasan_immunocomp/datasets/hartwig/biomarkers/database/drivers_full/somatic.txt\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
