{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "import os\n",
    "import gzip\n",
    "import pandas as pd \n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1",
   "metadata": {},
   "outputs": [],
   "source": [
    "sys.path.insert(0, f'{os.path.dirname(os.getcwd())}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2",
   "metadata": {},
   "outputs": [],
   "source": [
    "from map import I_DIR, SOMATIC_DIR, O_DIR\n",
    "#SOMATIC_DIR = I_DIR + 'somatics/'\n",
    "#O_DIR = I_DIR + 'biomarkers/database/'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3",
   "metadata": {},
   "source": [
    "### Helper Functions "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Functions to Read VCF file ###\n",
    "def fps(sample: str) -> str:\n",
    "    return SOMATIC_DIR + sample + \"/purple/\" + sample + \".purple.somatic.vcf.gz\"\n",
    "\n",
    "def read_vcf(vcf_path: str) -> pd.DataFrame:\n",
    "    names = get_vcf_col_names(vcf_path)\n",
    "    return pd.read_csv(vcf_path, compression='gzip', comment='#', chunksize=10000, delim_whitespace=True, header=None, names=names).read()\n",
    "\n",
    "def get_vcf_col_names(vcf_path: str) -> list:\n",
    "    with gzip.open(vcf_path, \"rt\") as ifile:\n",
    "          for line in ifile:\n",
    "            if line.startswith(\"#CHROM\"):\n",
    "                  vcf_names = [x for x in line.split('\\t')]\n",
    "                  break\n",
    "    ifile.close()\n",
    "    return vcf_names\n",
    "\n",
    "### Functions to extract essential fields for feature creation ###\n",
    "def get_metadata_df(vcf: pd.DataFrame) -> pd.DataFrame:\n",
    "    return vcf[[\"#CHROM\", \"POS\",  \"REF\", \"ALT\", \"FILTER\"]].rename(columns = {'#CHROM':'chromosome', \"POS\" : \"position\"})\n",
    "\n",
    "def filter_PASS( df: pd.DataFrame ) -> pd.DataFrame :\n",
    "    return df[[True if 'PASS' in i else False for i in df['FILTER']]]\n",
    "\n",
    "#### pass INFO column #### \n",
    "def get_impact_fields( s: str ) :\n",
    "    impact_fields = s.split(\"IMPACT=\")[1].split(\";\")[0].split(\",\")\n",
    "    return impact_fields    \n",
    "        \n",
    "def get_Gene(s: str):\n",
    "    if 'IMPACT=' in s: \n",
    "        return get_impact_fields( s )[0]\n",
    "    else:\n",
    "        return pd.NA \n",
    "\n",
    "def get_Transcript(s: str):\n",
    "    if 'IMPACT=' in s: \n",
    "        return get_impact_fields( s )[1]\n",
    "    else:\n",
    "        return pd.NA \n",
    "\n",
    "def get_Label(s: str):\n",
    "    if 'IMPACT=' in s: \n",
    "        return get_impact_fields( s )[2]\n",
    "    else:\n",
    "        return pd.NA    \n",
    "\n",
    "def get_field(s: str, f: str, n: bool) -> bool:\n",
    "    if f in s : \n",
    "        field = s.split(f)[-1].split(';')[0]\n",
    "        if n:\n",
    "            return float(field)\n",
    "        else: \n",
    "            return field\n",
    "    else:\n",
    "        return pd.NA\n",
    "\n",
    "def get_biallelic(s: str) -> bool:\n",
    "    if 'BIALLELIC' in s : \n",
    "        return True\n",
    "    else:\n",
    "        return False\n",
    "\n",
    "def extract_info_fields(s: str) -> dict:\n",
    "    return {\n",
    "        'gene': get_Gene(s), \n",
    "        'transcript' : get_Transcript(s), \n",
    "        'annotation': get_Label(s),\n",
    "        'purple_af': get_field(s, \"PURPLE_AF=\", True),\n",
    "        'purple_cn': get_field(s, \"PURPLE_CN=\", True),\n",
    "        'purple_vcn': get_field(s, \"PURPLE_VCN=\", True),\n",
    "        'purple_macn': get_field(s, \"PURPLE_MACN=\", True),\n",
    "        'tier': get_field(s, \"TIER=\", False),\n",
    "        'tnc' : get_field(s, \"TNC=\", False),        \n",
    "        'biallelic': get_biallelic(s),\n",
    "        'subclonal': get_field(s, \"SUBCL=\", True),\n",
    "     }    \n",
    "\n",
    "def get_info_df(vcf: pd.Series) -> pd.DataFrame:\n",
    "    return pd.DataFrame([extract_info_fields(i) for i in vcf['INFO']])\n",
    "    \n",
    "def vamos(sampleId: str) -> pd.DataFrame:\n",
    "    try:\n",
    "        fp = fps(sampleId)\n",
    "        if os.path.exists(fp):\n",
    "            vcf = read_vcf(fp)\n",
    "            metadata = get_metadata_df( vcf )\n",
    "            info = get_info_df( vcf )\n",
    "            ready = get_metadata_df(vcf).join(get_info_df(vcf))\n",
    "            ready = ready.query('FILTER==\"PASS\" & annotation == annotation & gene != \"\"')\n",
    "            #ready = ready.query('FILTER==\"PASS\" & (annotation == annotation | tier == \"HOTSPOT\") & (annotation not in (\"3_prime_UTR_variant\",\"5_prime_UTR_variant\",\"intron_variant\",\"upstream_gene_variant\",\"non_coding_transcript_exon_variant\"))')    \n",
    "            ready[['sampleId']] = sampleId\n",
    "        return ready\n",
    "    except Exception as e:\n",
    "        print(f\"Unexpected error: {e}\")\n",
    "        return None                 "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5",
   "metadata": {},
   "source": [
    "### Get file paths "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6",
   "metadata": {},
   "outputs": [],
   "source": [
    "files = os.listdir(SOMATIC_DIR)\n",
    "run_files = [i for i in files if \".\" not in i]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7",
   "metadata": {},
   "source": [
    "### Run and save output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8",
   "metadata": {},
   "outputs": [],
   "source": [
    "chunk_size = 702"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9",
   "metadata": {},
   "outputs": [],
   "source": [
    "dfs = []; k = 0; j = 0\n",
    "for i in run_files:\n",
    "    k = k + 1\n",
    "    print(k); print(i)\n",
    "    dfs.append(vamos(i))\n",
    "    if (k % chunk_size) == 0:\n",
    "        j = j + 1\n",
    "        pd.concat(dfs).to_csv(O_DIR + 'somatic_exome/somatic_exome_chunk_' + str(j) + '.csv', index = False)\n",
    "        dfs = []\n",
    "    if k == len(run_files):\n",
    "        j = j + 1\n",
    "        pd.concat(dfs).to_csv(O_DIR + 'somatic_exome/somatic_exome_chunk_' + str(j) + '.csv', index = False)    "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
