{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0",
   "metadata": {},
   "outputs": [],
   "source": [
    "source(paste0(dirname(getwd()),'/map.r'))\n",
    "source(paste0(HELP_DIR, \"shortcuts.r\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1",
   "metadata": {},
   "source": [
    "### 0 - Read and process collected driver data\n",
    "- Just including all drivers output from Linx here"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2",
   "metadata": {},
   "outputs": [],
   "source": [
    "som_exome <- fread(\"/mnt/petasan_immunocomp/datasets/hartwig/biomarkers/database/somatic_exome.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3",
   "metadata": {},
   "outputs": [],
   "source": [
    "#dnds <- readRDS(\"/mnt/petasan_immunocomp/datasets/hartwig/biomarkers/database/dnds/dnds.Rds\")\n",
    "sel_cvs <- fread(\"/mnt/petasan_immunocomp/datasets/hartwig/biomarkers/database/dnds/sel_cvs.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4",
   "metadata": {},
   "outputs": [],
   "source": [
    "purities <- fread(\"/mnt/petasan_immunocomp/datasets/hartwig/biomarkers/database/purities.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5",
   "metadata": {},
   "outputs": [],
   "source": [
    "cn_gene <- fread(\"/mnt/petasan_immunocomp/datasets/hartwig/biomarkers/database/cn_gene.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6",
   "metadata": {},
   "outputs": [],
   "source": [
    "breakends <- fread( paste0(TMP_DIR, \"structural_variants/breakend.csv\")) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7",
   "metadata": {},
   "outputs": [],
   "source": [
    "drivers <- fread( paste0(TMP_DIR, \"drivers.csv\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8",
   "metadata": {},
   "outputs": [],
   "source": [
    "#head(som_exome)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9",
   "metadata": {},
   "source": [
    "#### Check DNDS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "10",
   "metadata": {},
   "outputs": [],
   "source": [
    "tmp <- \n",
    "sel_cvs %>% \n",
    " mu(ct = n_syn + n_mis + n_non + n_spl + n_ind, \n",
    "    ratio = wnon_cv/wmis_cv, \n",
    "    type = ifelse(ratio > 1, \"TSG\", \"ONC\")) %>% \n",
    " fi(cohort == \"Pan-Cancer\") %>% \n",
    " se(cohort, gene_name, ct, type, ratio, contains(\"n_\"), contains(\"w\"), pglobal_cv, qglobal_cv) %>% \n",
    " ar(desc(ct))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "11",
   "metadata": {},
   "outputs": [],
   "source": [
    "tmp %>% ar(qglobal_cv) %>% fi(type == \"ONC\", grepl(\"FGF\", gene_name)) %>% ar(ratio)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "12",
   "metadata": {},
   "outputs": [],
   "source": [
    "pan <- tmp %>% fi(cohort == \"Pan-Cancer\") %>% pu(gene_name) %>% unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "13",
   "metadata": {},
   "outputs": [],
   "source": [
    "sel_cvs %>% \n",
    " fi(gene_name == \"BMP5\") %>% ar(pglobal_cv) %>% se(cohort, gene_name, contains(\"n_\"), contains(\"w\"), pglobal_cv) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "14",
   "metadata": {},
   "outputs": [],
   "source": [
    "tmp %>% fi(!gene_name %in% pan, ct > 10) %>% ar(pglobal_cv)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "15",
   "metadata": {},
   "outputs": [],
   "source": [
    "dnds$globaldnds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "16",
   "metadata": {},
   "outputs": [],
   "source": [
    "see <- fread(\"/mnt/petasan_immunocomp/datasets/hartwig/biomarkers/database/drivers.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "17",
   "metadata": {},
   "outputs": [],
   "source": [
    "drivers <- \n",
    "fread(\"/mnt/petasan_immunocomp/datasets/hartwig/biomarkers/database/drivers.csv\") %>% \n",
    " fi(likelihoodMethod == \"DNDS\") %>% \n",
    " gb(gene) %>% \n",
    " su(like = mean(driverLikelihood > .8), ct = n()) %>% mu(gene_name = gene) %>% \n",
    " ug()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "18",
   "metadata": {},
   "outputs": [],
   "source": [
    "see %>% \n",
    " gb(gene, likelihoodMethod) %>% \n",
    " su(ct = n()) %>% \n",
    " ar(desc(ct)) %>% \n",
    " fi(gene == \"PIK3CA\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "19",
   "metadata": {},
   "outputs": [],
   "source": [
    "dnds$sel_cv %>% ij(drivers, by = \"gene_name\") %>% se(gene_name, contains(\"n_\"), contains(\"w\"), like, ct, pglobal_cv) %>% gb(pglobal_cv < .01) %>% su(ct = n())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "20",
   "metadata": {},
   "outputs": [],
   "source": [
    "dnds$sel_cv %>% fi(qglobal_cv < .25) %>% fi(n_syn != 0)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "21",
   "metadata": {},
   "source": [
    "### Somatic Drivers "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "22",
   "metadata": {},
   "source": [
    "- Read in prepped exome file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "23",
   "metadata": {},
   "outputs": [],
   "source": [
    "somatic_exome <- \n",
    "som_exome %>% \n",
    " gb(chromosome, position, gene, tier) %>% \n",
    " mu(ct = n(), hotspot = (tier == \"HOTSPOT\" | ct > 30))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "24",
   "metadata": {},
   "source": [
    "- Hotspots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "25",
   "metadata": {},
   "outputs": [],
   "source": [
    "hotspots <- \n",
    "somatic_exome %>% \n",
    " fi(hotspot) %>% \n",
    " tm(sampleId, chromosome, gene, transcript, annotation, biallelic, method = \"HOTSPOT\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "26",
   "metadata": {},
   "source": [
    "- Biallelic"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "27",
   "metadata": {},
   "outputs": [],
   "source": [
    "biallelic <- \n",
    "somatic_exome %>% \n",
    " fi(!hotspot, biallelic) %>% \n",
    " tm(sampleId, chromosome, gene, transcript, annotation, biallelic, method = \"BIALLELIC\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "28",
   "metadata": {},
   "source": [
    "- Inframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "29",
   "metadata": {},
   "outputs": [],
   "source": [
    "inframe <- \n",
    "somatic_exome %>% \n",
    " fi(!hotspot, !biallelic, annotation %in% c(\"inframe_deletion\", \"inframe_insertion\", \"phased_inframe_insertion\", \"phased_inframe_deletion\")) %>% \n",
    " tm(sampleId, chromosome, gene, transcript, annotation, biallelic, method = \"INFRAME\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "30",
   "metadata": {},
   "source": [
    "- Somatic drivers "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "31",
   "metadata": {},
   "outputs": [],
   "source": [
    "som_drivers <- rbind(hotspots, biallelic, inframe) %>% lj(cn_gene %>% se(gene, chromosomeBand), by = \"gene\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "32",
   "metadata": {},
   "outputs": [],
   "source": [
    "#head(som_drivers)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "33",
   "metadata": {},
   "source": [
    "- Now DNDS! "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "34",
   "metadata": {},
   "outputs": [],
   "source": [
    "#dnds "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "35",
   "metadata": {},
   "source": [
    "### Copy number drivers "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "36",
   "metadata": {},
   "outputs": [],
   "source": [
    "amps <- \n",
    "cn_gene %>% \n",
    " ij(purities %>% se(sampleId, ploidy), by = \"sampleId\") %>%\n",
    " fi(maxCopyNumber > (3*ploidy)) %>% \n",
    " mu( method = ifelse(minCopyNumber > 3*ploidy, \"AMP\", \"PARTIAL_AMP\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "37",
   "metadata": {},
   "outputs": [],
   "source": [
    "dels <- \n",
    "cn_gene %>% \n",
    " ij( purities %>% se(sampleId, ploidy), by = \"sampleId\" ) %>%\n",
    " fi( minCopyNumber < .5 ) %>% \n",
    " mu( method = \"DEL\" )  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "38",
   "metadata": {},
   "outputs": [],
   "source": [
    "length(unique(cn_gene$sampleId))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "39",
   "metadata": {},
   "outputs": [],
   "source": [
    "head(dels)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "40",
   "metadata": {},
   "outputs": [],
   "source": [
    "drivers %>% fi(grepl(\"TRAF3\", gene))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "41",
   "metadata": {},
   "outputs": [],
   "source": [
    "dels %>% gb(chromosome, chromosomeBand, gene) %>% su(ct = n()) %>% ar(desc(ct)) %>% fi(chromosome == 23)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "42",
   "metadata": {},
   "source": [
    "### Structual Variant Drivers"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "43",
   "metadata": {},
   "source": [
    "- Disruptions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "44",
   "metadata": {},
   "outputs": [],
   "source": [
    "disruptions <- \n",
    "breakends %>%\n",
    " fi(disruptive) %>% \n",
    " tm(sampleId, chromosome, chromosomeBand = chrBand, gene, transcriptId, canonical, disruptive, method = \"DISRUPTION\") %>% \n",
    " unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "45",
   "metadata": {},
   "outputs": [],
   "source": [
    "head(disruptions)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "46",
   "metadata": {},
   "outputs": [],
   "source": [
    "disruptions %>% \n",
    " gb(chromosome, chromosomeBand, gene) %>% \n",
    " su(ct = n()) %>% \n",
    " ar(desc(ct))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
